packages/common/async/src/mutex.ts

```diff
@@ -133,3 +140,31 @@ export const synchronized = (
   };
   Object.defineProperty(descriptor.value, 'name', { value: propertyName + '$synchronized' });
 };
+
+export const acquireInContext = async (
+  ctx: Context,
+  mutex: Mutex,
+  options: { releaseOnDispose: boolean } = { releaseOnDispose: true },
+): Promise<MutexGuard> => {
+  let guard: MutexGuard | undefined;
+  try {
+    return await cancelWithContext(
+      ctx,
+      (async () => {
+        guard = await mutex.acquire();
+        if (ctx.disposed) {
+          guard.release();
+          guard = undefined;
+          throw new Error('Guard released because acquired after context was disposed.');
+        }
+        if (options.releaseOnDispose) {
+          ctx.onDispose(() => guard?.release());
+        }
+        return guard;
+      })(),
+    );
+  } catch (e) {
```

Comment:
`await cancelWithContext` throws if context is disposed, but yeah, the guard-acquiring promise should be enough:
```ts
        guard = await mutex.acquire();
        // if we're the second to settle we'll go inside this block and release
        // if we're the first to settle cancelWithContext won't throw 
        if (ctx.disposed) {
          guard.release();
          guard = undefined;
          throw new Error('Guard released because acquired after context was disposed.');
        }
        if (options.releaseOnDispose) {
          ctx.onDispose(() => guard?.release());
        }
        return guard;
```